### ISSUE TITLE
[vcpkg] Dumping all headers in the same include folder breaks encapsulation of CMake targets

### ISSUE THREAD

**Issue by petamas** [microsoft/vcpkg#38844]
**Describe the bug**

Most CMake projects simply set their include path to `${CMAKE_INSTALL_PREFIX}/include`. Since vcpkg installs all packages into the same prefix (`VCPKG_INSTALLED_DIR`), all the headers end up in various subdirectories of `${VCPKG_INSTALLED_DIR}/include`. However, this breaks encapsulation of the CMake targets: since the targets expose the same include folder as a usage requirement, depending on one of the vcpkg-based libraries from a CMake target gives access to the headers of ALL vcpkg-based libraries.

Allow me to demonstrate. (Note: the is issue is not specific to `fmt` and `units`, almost any two packages would expose the same issue; I chose these because they're both fairly small, and `units` is header-only, making the issue more subtle, and possibly more problematic.)

> `vcpkg.json`:
> ```
> {
>     "$schema": "https://raw.githubusercontent.com/microsoft/vcpkg-tool/main/docs/vcpkg.schema.json",
>     "dependencies": [
>         { "name": "fmt", "version>=": "8.1.1#2" },
>         { "name": "units", "version>=": "2.3.3#1" }
>     ],
>     "builtin-baseline": "943c5ef1c8f6b5e6ced092b242c8299caae2ff01"
> }
> ```
> 
> `CMakeLists.txt`:
> ```
> cmake_minimum_required (VERSION 3.20)
> 
> project (Demo)
> 
> set (CMAKE_CXX_STANDARD 17)
> set (CMAKE_CXX_STANDARD_REQUIRED ON)
> 
> find_package(fmt CONFIG REQUIRED)
> # find_package(units CONFIG REQUIRED)
> 
> add_executable (demo demo.cpp)
> target_link_libraries (demo
>     PRIVATE
>         fmt::fmt
>         # units::units
> )
> ```
> `demo.cpp`:
> ```
> #include <fmt/core.h>
> #include <units.h> // in an ideal world, this would be a compilation error
> 
> using namespace units::literals;
> 
> int main() {
>     units::area::square_meter_t area = 10_m * 10_m;
>     fmt::print("The area is {}", units::unit_cast<double>(area));
> }
> ```

The project above builds and runs correctly: `demo.cpp` has access to `units.h`, and can use all the types and functions from the `units` library, even though neither `fmt` or `demo` is depending on the `units::units` target, either directly or indirectly.

Naturally, if `units` wasn't a header-only library, we'd get link errors, but linking happens at a much later point in the build process than compilation, where the failure should happen IMO.

Even if we accept that the errors surface only at link time, there could be other problems: if `units` exposed other usage requirements in addition to the include folder, eg. macro definitions (using `target_compile_definitions`), or required compiler switches (using `target_compile_options`), those would not be applied while building `demo.cpp`, possibly causing errors. (Which may be possibly silent errors, or cause problems down the line.)

**Environment**
- OS: Windows 10 (but the same is true anywhere)
- Compiler: MSVC, cl.exe 19.38.33133.0 (but it doesn't matter)

**To Reproduce**
1. `cmake -GNinja -DCMAKE_BUILD_TYPE=Release -DCMAKE_TOOLCHAIN_FILE=path/to/vcpkg/scrips/buildsystems/vcpkg.cmake` path/to/source/folder
2. `ninja`

**Expected behavior**
The build should fail at the `#include <units.h>` line.

**Proposed solution**
I think it would be great if `vcpkg` automatically moved contents of the `include` folder to a package-specific subdirectory (eg. `include/${PORT}`), and the `vcpkg_cmake_config_fixup()` and similar functions updated the include paths to point into that subfolder, similarly how the CMake config files are handled.

**Comment by Neumann-A** [comment]
That is by design. a) MSBuild integration b) avoiding/detecting header collisions c) not really a vcpkg issues (feel free to complain upstream about their default installation layout)

**Comment by petamas** [comment]
@Neumann-A : Thanks for the quick reply! I didn't know whether it was by design or an oversight. I can accept that this behaviour is by design, but I'd like to react to some of the rationale.

Regarding c): I don't feel like this is a problem on the package authors' part - I think most libraries' CMakeLists.txt file assumes that the install prefix is library-specific, i.e. only their files end up there. I'm not sure whether that's a good assumption or not, but I don't think it's economic to tackle this package-by-package if it can be fixed semi-automatically for most packages in vcpkg's catalog.

Regarding b): I don't think putting everything in the same folder avoids header collisions; there can be both false negatives (i.e. collisions tat are not detected) and false positives.

My understanding of "header collision" is when an `#include <something.h>` statement in a client is ambigious, i.e. when two different packages provide `<something.h>`.

False negative example:
- Package `A` adds `include/foo/bar.h`, and exposes `include/foo` as an include directory. 
- Package `B` adds `include/bar.h`, and exposes `include` as an include directory.
- Since the two `bar.h` files are copied into separate places within `vcpkg_installed`, no collision is detected, but `include<bar.h>` is still ambiguous.
- This case is neither fixed nor worsened by my proposed change.

False positive example:
- Package `A` adds `include/foo/bar.h`, and exposes `include` as an include directory. 
- Package `B` adds `include/foo/bar.h`, and exposes `include/foo` as an include directory.
- Since the two `bar.h` files are copied into the same place `vcpkg_installed`, a collision is detected.
- OTOH, if my proposal was implemented, the files would be copied to separate places (`include/A/foo/bar.h` and `include/B/foo/bar.h`), and there would be no collisions: `#include<foo/bar.h>` would unambiguously refer to `A`'s version, while `#include<bar.h>` would unambiguously refer to `B`'s version, just as the package authors intended.

In conclusion, I still feel like it would be beneficial to solve this from the vcpkg side (either by extending `vcpkg_cmake_config_fixup()`, or introducing a similar helper function that can be used by portfiles. However, I can work around this by providing a custom overlay port of `vcpkg-cmake-config` in my project, so feel free to close this issue if you don't want to pursue this direction.
